rom/laghos.cpp

```diff
@@ -1292,27 +1405,41 @@ int main(int argc, char *argv[])
                 // TODO: it needs to be save in the format of HDF5 format
                 // TODO: how about parallel version? introduce rank in filename
                 // TODO: think about how to reuse "gfprint" option
-                std::string filename = outputPath + "/ROMsol/romS_" + std::to_string(ti);
-                std::ofstream outfile_romS(filename.c_str());
-                outfile_romS.precision(16);
-                if (romOptions.hyperreduce && romOptions.GramSchmidt)
-                {
-                    Vector romCoord(romS);
-                    romOper[romOptions.window]->InducedGramSchmidtFinalize(romCoord, true);
-                    romCoord.Print(outfile_romS, 1);
-                }
-                else
+                if (!rom_build_database)
                 {
-                    romS.Print(outfile_romS, 1);
+                    std::string filename = outputPath + "/ROMsol/romS_" + std::to_string(ti);
+                    std::ofstream outfile_romS(filename.c_str());
+                    outfile_romS.precision(16);
+                    if (romOptions.hyperreduce && romOptions.GramSchmidt)
+                    {
+                        Vector romCoord(romS);
+                        romOper[romOptions.window]->InducedGramSchmidtFinalize(romCoord, true);
+                        romCoord.Print(outfile_romS, 1);
+                    }
+                    else
+                    {
+                        romS.Print(outfile_romS, 1);
+                    }
+                    outfile_romS.close();
                 }
-                outfile_romS.close();
 
                 if (!romOptions.hyperreduce)
-                    basis[romOptions.window]->LiftROMtoFOM(romS, *S);
+                {
+                    if (!rom_build_database || last_step)
+                    {
+                        basis[romOptions.window]->LiftROMtoFOM(romS, *S);
+                    }
+                    if (rom_build_database && last_step)
+                    {
+                        lastLiftedSolution = *S;
+                        ode_solver_dat->Init(*oper);
+                        ode_solver_dat->Step(lastLiftedSolution, t, dt);
```

Comment:
At last step, initialize the ode_solver_dat and take a single step in the FOM space. This will be used for the residual.

Reply:
I'm not sure, but it may be possible to use `ode_solver` here, if it is not used later. That is, you can change the operator and solution state for an `ODESolver`. It may be worth trying, and if it doesn't work, it's not a big deal.